### ISSUE TITLE
DD gets stuck non-deterministically

### ISSUE THREAD

**Issue by ryzhyk** [TimelyDataflow/differential-dataflow#223]
The following trivial program simply copies an input collection to an output one. We start 100 workers. The first worker advances the timestamp without feeding any actual data to the pipeline. Other workers simply drop their input handles.

The important part here is the barrier that forces all workers to synchronize before starting a new round (I cannot reproduce the failure without it).

On my laptop this usually runs for <5000 rounds before all workers get stuck (using the current DD master).

Am I using the DD API incorrectly or is this indeed a bug in DD?

```rust
extern crate rand;
extern crate timely;
extern crate differential_dataflow;

use timely::dataflow::operators::probe::Handle;

use differential_dataflow::input::Input;
use differential_dataflow::Collection;
use differential_dataflow::operators::*;
use timely::communication::initialize::Configuration;

use std::sync::{Arc,Barrier};

fn main() {
    let nworkers = 100;

    let progress_barrier = Arc::new(Barrier::new(nworkers));

    // define a new computational scope, in which to run BFS
    timely::execute(Configuration::Process(nworkers), move |worker| {
        let progress_barrier = progress_barrier.clone();

        let mut probe = Handle::new();
        let mut edge = worker.dataflow(|scope| {

            let (edge_input, edge): (_, Collection<_, (usize, usize), isize>) =  scope.new_collection();

            let result = edge;

            result.consolidate()
                  .probe_with(&mut probe);

            edge_input
        });

        if worker.index() == 0 {
            for i in 2..1_000_000_000 {
                edge.advance_to(i);
                edge.flush();

                eprintln!("i: {}", i);
                progress_barrier.wait();
                while probe.less_than(&i) {
                    worker.step_or_park(None);
                }
            }
        } else {
            edge.close();
            for i in 2..1_000_000_000 {

                progress_barrier.wait();
                while probe.less_than(&i) {
                    worker.step_or_park(None);
                }
            }
        }
        eprintln!("done");
    }).unwrap();
}
```

This was originally discovered by @ekcs.

@mbudiu-vmw

**Comment by ryzhyk** [comment]
Upd: according to `gdb`, when the program is stuck all workers are waiting on the barrier, except for one that is stuck inside `step_or_park`. Here is this guy's stack trace:

```
(gdb) bt
#0  0x00007f909a43f9f3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x55a480eada6c)
    at ../sysdeps/unix/sysv/linux/futex-internal.h:88
#1  __pthread_cond_wait_common (abstime=0x0, mutex=0x55a480eada10, cond=0x55a480eada40) at pthread_cond_wait.c:502
#2  __pthread_cond_wait (cond=0x55a480eada40, mutex=0x55a480eada10) at pthread_cond_wait.c:655
#3  0x000055a47eee7573 in std::sys::unix::condvar::Condvar::wait () at src/libstd/sys/unix/condvar.rs:71
#4  std::sys_common::condvar::Condvar::wait () at src/libstd/sys_common/condvar.rs:41
#5  std::sync::condvar::Condvar::wait () at src/libstd/sync/condvar.rs:204
#6  std::thread::park () at src/libstd/thread/mod.rs:911
#7  0x000055a47ee43324 in <timely_communication::allocator::thread::Thread as timely_communication::allocator::Allocate>::await_events (self=0x7f903c002530, 
    duration=...) at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/communication/src/allocator/thread.rs:44
#8  0x000055a47ee57216 in <timely_communication::allocator::process::Process as timely_communication::allocator::Allocate>::await_events (
    self=0x7f903c002530, duration=...)
    at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/communication/src/allocator/process.rs:182
#9  0x000055a47ee438f8 in <timely_communication::allocator::generic::Generic as timely_communication::allocator::Allocate>::await_events (
    self=0x7f903c002528, _duration=...)
    at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/communication/src/allocator/generic.rs:100
#10 0x000055a47ec0fc78 in timely::worker::Worker<A>::step_or_park (self=0x7f90735f8bf8, duration=...)
    at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/timely/src/worker.rs:209
#11 0x000055a47ed37b91 in bug::main::{{closure}} (worker=0x7f90735f8bf8) at examples/bug.rs:53
#12 0x000055a47ecbb434 in timely::execute::execute::{{closure}} (allocator=...)
    at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/timely/src/execute.rs:201
#13 0x000055a47ed2ec3b in timely_communication::initialize::initialize_from::{{closure}} ()
    at /home/lryzhyk/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/f2b0cc7/communication/src/initialize.rs:269
#14 0x000055a47ec3f063 in std::sys_common::backtrace::__rust_begin_short_backtrace (f=...)
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/sys_common/backtrace.rs:77
#15 0x000055a47ec1e255 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} ()
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/thread/mod.rs:470
#16 0x000055a47ed2efd5 in <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=())
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panic.rs:315
#17 0x000055a47ecc235c in std::panicking::try::do_call (data=0x7f90735f97d0 "\001")
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panicking.rs:296
#18 0x000055a47eef561a in __rust_maybe_catch_panic () at src/libpanic_unwind/lib.rs:80
#19 0x000055a47ecc21b9 in std::panicking::try (f=...) at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panicking.rs:275
#20 0x000055a47ed2f015 in std::panic::catch_unwind (f=...) at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/panic.rs:394
#21 0x000055a47ec1e039 in std::thread::Builder::spawn_unchecked::{{closure}} ()
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libstd/thread/mod.rs:469
#22 0x000055a47ecd0743 in core::ops::function::FnOnce::call_once{{vtable-shim}} ()
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/libcore/ops/function.rs:235
---Type <return> to continue, or q <return> to quit---
#23 0x000055a47eee678f in <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once ()
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/liballoc/boxed.rs:787
#24 0x000055a47eef4d50 in <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once ()
    at /rustc/625451e376bb2e5283fc4741caa0a3e8a2ca4d54/src/liballoc/boxed.rs:787
#25 std::sys_common::thread::start_thread () at src/libstd/sys_common/thread.rs:13
#26 std::sys::unix::thread::Thread::new::thread_start () at src/libstd/sys/unix/thread.rs:79
#27 0x00007f909a4396db in start_thread (arg=0x7f90735fa700) at pthread_create.c:463
#28 0x00007f9099f4a88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Comment by frankmcsherry** [comment]
It seems to hang with just two threads also. I'm not sure if it's a bug; interleaving concurrency mechanisms is definitely a scary thing that doesn't always work out the way you want. While it may be a bug, it is important to ack that the `while` loops are not imperative execution, they are reactive and awaiting others.

**Comment by ryzhyk** [comment]
Understood, but once the `while` loop condition is satisfied, doesn't this guarantee that the worker has finished all work for the given timestamp and should not block other workers from progressing to the same timestamp?

**Comment by frankmcsherry** [comment]
So for example, I'm not sure that anything about timely or differential guarantees that if you call `edge.flush()` and then go in to a barrier that you should expect other people to receive the results of your flush and wake you out of the barrier.

**Comment by frankmcsherry** [comment]
Generally, calling `flush()` will provoke dataflow work to do, which will lead to progress messages being communicated *if you repeatedly call `worker.step()`*. The only thing that has happened after `edge.flush()` is that an input operator has some progress updates it is ready to send, but no one is going to get to them without a call to `worker.step()` and if thread zero goes and blocks on a barrier everyone is just going to wait.

**Comment by frankmcsherry** [comment]
if it is at all helpful, timely has [its own barrier construct](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/synchronization/barrier.rs). The main difference is that the thread completes work and then remains in a wake-able state in case input messages arrive, unlike a system-provided barrier which prevents the thread from doing or reacting to work.

**Comment by frankmcsherry** [comment]
Though, now that I read it it was written without `step_or_park()` smarts and will just busy-wait, which I bet you don't want. But I could imagine we could fix that pretty easily...

**Comment by ryzhyk** [comment]
> So for example, I'm not sure that anything about timely or differential guarantees that if you call edge.flush() and then go in to a barrier that you should expect other people to receive the results of your flush and wake you out of the barrier.

I don't think my code makes such an assumption. The barrier only waits for workers to finish what they were doing for the previous timestamp (`i-1`).

**Comment by frankmcsherry** [comment]
It's hard for me to reason about possible interleavings here, but I don't see an obvious argument that all is well.

Another issue is that timely uses thread `park()` and `unpark()` machinery to wake threads. I imagine that if a thread is in a barrier and receives an `unpark()` it absorbs the call and prevents exciting the barrier due to a "spurious" wake-up. I suspect there are negative implications of this, but I can't really tell what they are.

Is the independent barrier an important feature of your program?

**Comment by ryzhyk** [comment]
I can see your point about barriers potentially not playing well with the `park()/unpark()`-based scheduling. I am keen to switch to a more DD-compatible synchronization method (will post a separate comment about it), but I think it would still be good to understand whether in this particular case barriers cause the problem or just help uncover an unrelated bug. I actually over-simplified the example slightly before submitting the issue. Below is a slightly modified version where I moved the barriers to after the `while` loops.

Here is how I attempt to convince myself that this code is correct:
- by the time a worker reaches the barrier, it is done with all processing for the current timestamps and will not receive any more wake-ups for it, so it is safe to sleep
- when the last worker (let's call it `W`) reaches the barrier, it starts waking up everyone else.  This is where things get a bit funny. As a worker wakes up, it starts processing updates, potentially calling `unpark` on other workers.  It is possible that some of these workers still haven't been woken up, and so the barrier will absorb those `unpark()`s, but this should not be a problem, as `W` will wake up those workers eventually, at which point they will have a chance to observe whatever changes other workers sent their way.

Does this make sense? 

```rust
extern crate rand;
extern crate timely;
extern crate differential_dataflow;

use timely::dataflow::operators::probe::Handle;

use differential_dataflow::input::Input;
use differential_dataflow::Collection;
use differential_dataflow::operators::*;
use timely::communication::initialize::Configuration;

use std::sync::{Arc,Barrier};

fn main() {
    let nworkers = 100;

    let progress_barrier = Arc::new(Barrier::new(nworkers));

    // define a new computational scope, in which to run BFS
    timely::execute(Configuration::Process(nworkers), move |worker| {
        let progress_barrier = progress_barrier.clone();

        let mut probe = Handle::new();
        let mut edge = worker.dataflow(|scope| {

            let (edge_input, edge): (_, Collection<_, (usize, usize), isize>) =  scope.new_collection();

            let result = edge;

            result.consolidate()
                  .probe_with(&mut probe);

            edge_input
        });

        if worker.index() == 0 {
            for i in 2..1_000_000_000 {
                edge.advance_to(i);
                edge.flush();

                eprintln!("i: {}", i);
                while probe.less_than(&i) {
                    worker.step_or_park(None);
                }
                progress_barrier.wait();
            }
        } else {
            edge.close();
            for i in 2..1_000_000_000 {

                while probe.less_than(&i) {
                    worker.step_or_park(None);
                }
                progress_barrier.wait();
            }
        }
        eprintln!("done");
    }).unwrap();
}
```

**Comment by frankmcsherry** [comment]
We can investigate, but .. it's a bit speculative ("look for bad interleaving; trying to sort out if it is a bug or an unexpected interaction"). There could easily be places, for example, where the worker defers checking input channels before a `park()` on the premise that if there is data then its thread token should be available and the `park()` would immediately return. This would be "wrong" if someone else is able to eat the `unpark()` calls and not leave the thread's token available, but isn't (afaiu) wrong otherwise.

**Comment by frankmcsherry** [comment]
But, the above case does seem like a sane thing to audit for (at the moment): at the entry to `step_or_park()` we could assume that the thread token is not available, because who knows what mayhem has happened, and the worker should take care to perform as many tests as possible *before* calling `park()`.

I can't think of a good reason *not* to do this, but it will take some looking to make sure this happens (it's a timely dataflow issue, in that case).

**Comment by ryzhyk** [comment]
> There could easily be places, for example, where the worker defers checking input channels before a park() on the premise that if there is data then its thread token should be available and the park() would immediately return.

Ok, then using `step_or_park` with `sync::Barrier` is clearly not a good idea. I will think about a different scheme, e.g., using timely's barrier.  But now I'm also worried about the effect of blocking operations inside the dataflow itself. E.g., what if `inspect` calls, e.g., `println!()`, which happens to block on I/O (I don't know that it does, just guessing). Can this lead to the same situation you describe above?

**Comment by frankmcsherry** [comment]
That's a great question! I'm not sure what happens, and it seems reasonable to be worried about it!

**Comment by frankmcsherry** [comment]
Quick inspection of the code suggests that each call to `Worker::step_or_park()` 

1. first calls the communication layer's `receive()` method, which populates the communication layer's `events` queue (and which prevents parking if non-empty),
2. then advances the list of dataflow activations (capturing all that are available at the moment), and sets the park duration to zero if any activations exists.

Only after these two things have happened, and only if the park duration is non-zero does it call in to `await_events()` which is the potentially parking code in the communication layer. It does not appear that there is a problem with entering `step_or_park()` without the thread token, in that assuming that any signaled work is actually visible at this point (e.g. available to the communication plane) it should be surfaced by `receive` and then prevent parking. If the work was produced locally, it should be visible in activations and also prevent parking.

So, not sure what to say at the moment. If you end up with a read-out of the state of the workers in the deadlocked state I'd be interested, but "state" in this case would be the contents of various activation queues and progress buffers and such. It's possibly more than you want to do. I can put it on my queue but .. that queue is growing.

**Comment by ryzhyk** [comment]
Thanks for the explanation!  So, we're back to "barriers are not pretty, but we don't have a clear explanation for how they cause the pipeline to get stuck, which is concerning"  state. I will try to investigate what the worker's state is and how it got there...

**Comment by ryzhyk** [comment]
I think there is a race condition between these two lines:

https://github.com/TimelyDataflow/timely-dataflow/blob/a901170e6610406c69c077501ec4f9e5f60ac7a1/communication/src/allocator/counters.rs#L95

and

https://github.com/TimelyDataflow/timely-dataflow/blob/a901170e6610406c69c077501ec4f9e5f60ac7a1/communication/src/allocator/counters.rs#L100

They send two mpsc messages. The receiver can process the first message (`Push()`) and park without receiving the second message.  Even though the second message is followed by `unpark()`, which will wake the receiver up, the receiver may still never get around to receiving this message, as it first checks for outstanding `Push/Pull` messages and only creates an activation if there is one. But since we drained these messages in the previous iteration of `step_or_park`, they are basically lost.

```
worker1                                            worker2
   |                                                |-enters step_or_park()
   |                                                |
   |-self.events.send(Event::Pushed)--------------->|-receives Event::Pushed
   |                                                |-calls step() on the dataflow that this event belongs to
   |                                                |-step() returns
   |                                                |-step_or_park() returns
   |                                                |-enters step_or_park()
   |                                                |-nothing to do; park()
   |-self.pusher.push(element)--------------------->|-wakes up, reenters step_or_park()
   |     (unparks worker2)                          |-no new events; park()
```

**Comment by frankmcsherry** [comment]
Interesting! I'll look into it. The original intent (from the comments) is that this is actually used for communication within the same thread, just when it needs to drop things back in a `Send` based queue. It is totally possible that the comment is wrong and the design shifted and it is now used differently, and I'll take a look and see what I can find!

Thanks very much for looking around!

**Comment by ryzhyk** [comment]
I think that comment was copied from `struct Pusher`. I am fairly sure `ArcPusher` is "intra-process", not "intra-thread".

**Comment by frankmcsherry** [comment]
It's definitely in the wrong order (afaict), and looking quickly does imply that it is used between processes (in `process.rs`, the intra-process, inter-thread allocator). I'll put together a fork and let you check it out (and will also check it out as I can, but .. have a flight coming up).

**Comment by frankmcsherry** [comment]
I put up a timely branch (and cc'd you). I'm still not 100% clear, as my understanding of the counter that goes in is that it increases a counter that prevents the remote worker from winding down (a count of the number of un-received communications). Once it receives that count, it knows something is coming in the channel and should not sleep. It is possible this is not how it works, and I need to get my head back around this again, sorry!

**Comment by frankmcsherry** [comment]
One conclusion is that I should write down the concurrency discipline clearly, and subject it to public scrutiny / embarrassment.

**Comment by ryzhyk** [comment]
I don't think reordering solves the problem. As far as I can tell, `step_or_park` ignores the counter.

No rush with the fix. I realize it is a subtle problem and requires time to solve properly. Unfortunately, my understanding of the code is nowhere near good enough to come up with my own fix.

**Comment by frankmcsherry** [comment]
I filed a timely issue for this. It seems to be not specific to differential dataflow, and I'll try and figure out what is going on over there.

**Comment by frankmcsherry** [comment]
Should be closed by https://github.com/TimelyDataflow/timely-dataflow/pull/298, but please re-open if that turns out to be premature.
